### DISCUSSION TITLE
HTTP API get '/queues' response delay up to 40+seconds when 1 of 3 cluster nodes goes down (quorum queues)

### DISCUSSION THREAD

**Discussion by fare1990** [rabbitmq/rabbitmq-server#6048] [Other] [answered]
RabbitMQ 3.10.6, Erlang 23.3.4.7. 3-node cluster, quorum queues.

I'm faced with the issue using RabbitMQ cluster (3 nodes), quorum queues & management-plugin's REST API.

In conditions when all nodes are up & running GET request for queues list (GET /api/queues) takes at most 300ms.

If we shutdown one of cluster node completely (with OS) - the same request started to hang for up to 40 seconds. 

After short investigation figured out:
1) Only HTTP  'api/queues'  endpoint affected, AMQP operations continue to work well. Looks like other HTTP endpoints that do not perform queue metrics collection working well too.
2) Issue happens when EPMD port 4369 became unreachable on specific node for other cluster member nodes.
3) Issue does not reproduce when classic queues mirroring used, only with quorum queues.
4) Playing with configuration flags to disable metrics gathering & with tick time - no effect.

**Steps to reproduce:**
1) Configure cluster with 3 nodes, enable management plugin for all.
Config example:
```
cluster_name                = somename
cluster_partition_handling  = pause_minority
cluster_formation.node_type = disc

cluster_formation.classic_config.nodes.1 = [rabbit@zmpha-wfan-10024-master1.wflab.io](mailto:rabbit@zmpha-wfan-10024-master1.wflab.io)
cluster_formation.classic_config.nodes.2 = [rabbit@zmpha-wfan-10024-master2.wflab.io](mailto:rabbit@zmpha-wfan-10024-master2.wflab.io)
cluster_formation.classic_config.nodes.3 = [rabbit@zmpha-wfan-10024-master3.wflab.io](mailto:rabbit@zmpha-wfan-10024-master3.wflab.io)

ssl_options.cacertfile           = /somefolder/ssl/ca.crt
ssl_options.certfile             =  /somefolder/ssl/server.crt
ssl_options.keyfile              =  /somefolder/ssl/server.key
ssl_options.verify               = verify_peer
ssl_options.fail_if_no_peer_cert = true

ssl_options.versions.1 = tlsv1.3
ssl_options.ciphers.1  = TLS_AES_256_GCM_SHA384
ssl_options.ciphers.2  = TLS_AES_128_GCM_SHA256
ssl_options.ciphers.3  = TLS_CHACHA20_POLY1305_SHA256
ssl_options.ciphers.4  = TLS_AES_128_CCM_SHA256
ssl_options.ciphers.5  = TLS_AES_128_CCM_8_SHA256

vm_memory_high_watermark.absolute = 2048MiB

net_ticktime = 60
consumer_timeout = 43200000

management.tcp.port = 15672
management.tcp.ip   = 127.0.0.1
```
`RABBITMQ_USE_LONGNAME` was set to true.
2) Create some number of durable Quorum queues, for example by setting `x-queue-type=quorum` flag.

3)Stop OS on one of the node ([zmpha-wfan-10024-master1.wflab.io](http://zmpha-wfan-10024-master1.wflab.io/)  for example), perform HTTP GET request to any other node.
**Alternative**: you can stop RabbitMQ server on the node (everything still works well) and then block traffic on this node for port 4369 (`iptables -A INPUT -p tcp -s zmpha-wfan-10024-master2.wflab.io --dport 4369 -j DROP && iptables -A INPUT -p tcp -s zmpha-wfan-10024-master3.wflab.io --dport 4369 -j DROP`) - after this REST API calls started to hang.

-----------------------------------------------------------------------------

This behavior block some pretty standard maintenance operations, cause in our scenario we need to periodically gather queue statistics.

**Comment by michaelklishin** [comment, marked as answer]
This works expected.

When you query all queues or all connections or other "all things" endpoints, the node that handles the request aggregates the results from its peers and returns them to the client. There is a certain timeout involved. If one node is disconnected without being shut down, all requests to it from its peers will block until they time out.

Monitoring using `GET /api/queues`, `GET /api/connections` and in late 2022, using HTTP API queries at all is wrong. It is very common to see people use those endpoints to get a single field from a single object. That's really wasteful, as most of the metrics returned are not used at all.

This problem is not present in the [Prometheus endpoint](https://www.rabbitmq.com/prometheus.html) because every node returns only its own stats, and the aggregation is done at a later point by tools such as Grafana. Prometheus endpoing scraping is recommended for other reasons which are documented in the [RabbitMQ guide on monitoring](https://www.rabbitmq.com/monitoring.html).

**Reply by fare1990** [reply]
This does not reproduced when you use 'classic' queues, only for quorum.
When we shutdown node completely the issue still reproduced.
It looks really weird to have API endpoint that reacts differently depending on the queue types you use. Also it's not clear why we need to wait for connection timeout from dead node while gathering some data instead of ignoring this node in request (information that this node is not available can be gathered, even 'overview' UI tab which I think use the same REST API responses without any delays).

We use additional information about queues (for example consumer count, queue arguments list, idle since, acknowledged messages count etc.) from all queues in our brokering mechanism. Unfortunately this info could not be gathered using AMQP.

**Reply by lukebakken** [reply]
My guess is that with classic queues the queue processes are not queried directly. There's probably improvements that can be made with quorum queues / ra in this scenario (@kjnilsson ?)

> We use additional information about queues (for example consumer count, queue arguments list, idle since, acknowledged messages count etc.) from all queues in our brokering mechanism. Unfortunately this info could not be gathered using AMQP.

As @michaelklishin said, you should look into getting as much of this information as possible from Prometheus.

**Reply by fare1990** [reply]
Unfortunately Prometheus metrics endpoint does not provide all necessary information and useless for us (
The only way I can get queue arguments is REST API.

**Reply by michaelklishin** [reply]
Then you have to accept this design limitations of the HTTP API: every node aggregates results from other nodes. If some of those nodes are not available, there will be a delay. The timeout may be configurable but if you query *all* queues then I would not try to lower it as you'd get false negative timeouts.

**Reply by michaelklishin** [reply]
Also, Prometheus does provide [per object metrics](https://www.rabbitmq.com/prometheus.html#metric-aggregation) but it is disable for the standard endpoint by default as it can produce gigantic responses that take a long time to render.

**Reply by kjnilsson** [reply]
How many queues are we talking about here?

**Reply by kjnilsson** [reply]
The only thing I could think is that there are many queues and grabbing metrics for all queues cause a connection attempt to the blocked port and this for some reason is taking long enough for this work to add up to many seconds of delay.

If so I am sure we can improve on that.

I did some testing with only a few quorum queues and could not see more than a few ms delay for the /queues endpoint but I didn't use TLS so we'd need more testing. It would be interesting to know if #https://github.com/rabbitmq/rabbitmq-server/pull/6038 improves matters at all.

**Reply by fare1990** [reply]
@kjnilsson @michaelklishin 
Thanks,
We have ~20 queues. 100% Reproduced for our two fresh environments.
We'll try to deal with the delay in such conditions, looking forward if some improvements will be made

**Reply by kjnilsson** [reply]
Are you able to try latest `main` version to see if it improves matters?

**Reply by fare1990** [reply]
3.11.0 you mean.
It requires erlang 25, I'm not sure we can test it quickly.
I'll try to test it a bit later, thanks.

**Reply by kjnilsson** [reply]
No I mean testing of the bleeding edge `main` branch. We produce docker images for each commit that can be used for testing. That said it appears the change I am referring to has been backported to the 3.10.x branch and thus can be run on 24.

here is a link to the image: https://hub.docker.com/layers/pivotalrabbitmq/rabbitmq/2005ba164a0b7245a34a924fcc15a4c511eb8be6-otp-min/images/sha256-31c394e63bae29d3be5203475dc08e37af35c2f5b21186ab3e5d9904a082998f?context=explore

**Reply by fare1990** [reply]
To test it exactly in the same conditions I need to build rpm and install it to our AWS environment.

**Reply by kjnilsson** [reply]
https://github.com/rabbitmq/rabbitmq-server/releases/tag/v3.10.9

This release contains the change I mentioned

It may improve matters somewhat - it would be interesting to see either way.

**Reply by fare1990** [reply]
Cool, thanks. 
I will test it in few days

**Reply by fare1990** [reply]
Tested - no changes, still faced with delays.
Currently we are going with temporary work-around by adding firewall rule for running nodes by rejecting outgoing EPMD port connection to stopped node. `iptables -A OUTPUT -p tcp -d zmpha-wfan-10024-master1.wflab.io --dport 4369 -j REJECT --reject-with tcp-reset` for example for both master2 & master3 if master1 goes down.
With this rule everything works without any delays. Interesting here is that `--reject-with tcp-reset` option matters, no delays only if we drop connection immediately.

**Reply by michaelklishin** [reply]
The best suggestion I have is using [maintenance mode](https://www.rabbitmq.com/upgrade.html#maintenance-mode), although it is not always easy to do in various scenarios on Kubernetes.

Then if nodes would filter out their peers under maintenance, you'd avoid hitting the nodes that are known to be temporarily stopped.

**Reply by kjnilsson** [reply]
@fare1990 the delays are as long as before?

**Reply by kjnilsson** [reply]
One thing that would be helpful for us is if you could identify if the delay is caused by the inclusion of certain columns in the GET request. So if you add e.g. `columns=name,state` etc. to your query string you should be able to identify which columns cause the delay and which don't. E.g. I would not expect a request for only the `name` to have undue delay.

**Reply by fare1990** [reply]
yeap, the delays are the same - increased linear with queues count.

I will experiment with columns, thanks.

**Reply by fare1990** [reply]
Even with only name column requested - same delays
![image](https://user-images.githubusercontent.com/18418615/196428383-af395c48-2ffb-4a13-b0e9-8033932500fc.png)

**Reply by kjnilsson** [reply]
Ok I can see now that there are some paths in the code that do fanout stat collection even if the fields aren't included.

I've started some tweaks in this PR: https://github.com/rabbitmq/rabbitmq-server/pull/6183

If you're at all able to test it in your environment it would of course be helpful and speed eventual resolution up.

**Reply by fare1990** [reply]
@kjnilsson Great, thanks!

I'm able to test it with rpm available

**Reply by fare1990** [reply]
@kjnilsson @michaelklishin 
I've tested 3.10.11 with improvement & it works perfectly. Thanks!

**Reply by kjnilsson** [reply]
Great thanks for letting us know.
